gemini/sample-apps/e2e-gen-ai-app-starter-pack/app/eval/utils.py

```diff
@@ -0,0 +1,244 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from concurrent.futures import ThreadPoolExecutor
+from functools import partial
+import glob
+import logging
+import os
+from typing import Any, Callable, Dict, Iterator, List
+
+import nest_asyncio
+import pandas as pd
+from tqdm import tqdm
+import yaml
+
+nest_asyncio.apply()
+
+
+def load_chats(path: str) -> List[Dict[str, Any]]:
+    """
+    Loads a list of chats from a directory or file.
+
+    Args:
+        path (str): The path to the directory or file containing the chats.
+
+    Returns:
+        List[Dict[str, Any]]: A list of chats.
+    """
+
+    chats: List[Dict[str, Any]] = []
+    for path in glob.glob(path):
+        with open(path) as f:
+            chats_in_file = yaml.safe_load(f)
+            chats = chats + chats_in_file
+    return chats
+
+
+def pairwise(iterable: List[Any]) -> Iterator[tuple[Any, Any]]:
+    """Creates an iterable with tuples paired together
+    e.g s -> (s0, s1), (s2, s3), (s4, s5), ...
+    """
+    a = iter(iterable)
+    return zip(a, a)
+
+
+def generate_multiturn_history(df: pd.DataFrame) -> pd.DataFrame:
+    """Processes a DataFrame of conversations to create a multi-turn history.
+
+    This function iterates through a DataFrame where each row represents a conversation.
+    It extracts human and AI messages from the "messages" column and structures them
+    into a new DataFrame. Each row in the output DataFrame represents a single turn
+    in a conversation, including the human message, AI message, and the conversation
+    history up to that point.
+
+    Args:
+        df (pd.DataFrame): A DataFrame where each row represents a conversation.
+                           The DataFrame should have a column named "messages" containing
+                           a list of alternating human and AI messages.
+
+    Returns:
+        pd.DataFrame: A DataFrame where each row represents a single turn in a conversation.
+                      The DataFrame has the following columns:
+                          - human_message: The human message in that turn.
+                          - ai_message: The AI message in that turn.
+                          - conversation_history: A list of all messages in the conversation
+                                                  up to and including the current turn.
+    """
+    processed_messages = []
+    for _, row in df.iterrows():
+        conversation_history: List[str] = []
+        for human_message, ai_message in pairwise(row["messages"]):
+            message = {
+                "human_message": human_message,
+                "ai_message": ai_message,
+                "conversation_history": conversation_history,
+            }
+            conversation_history = conversation_history + [human_message, ai_message]
+            processed_messages.append(message)
+
+    return pd.DataFrame(processed_messages)
```

Comment:
See if there's a way to change this to use dataframe operations, like `df.apply()` could make it more efficient.